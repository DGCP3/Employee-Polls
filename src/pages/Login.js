import { useState } from "react";
import { Navigate, useLocation } from "react-router-dom";
import { Button } from "../components/QuestionCard";
import useStore from "../hooks/useStore";
import { Form, Input } from "./NewPoll";

export default function Login() {
  const [form, setForm] = useState({
    username: "zoshikanlu",
    password: "pass246",
  });

  const {
    store: { isAuth, loading },
    loginThunk,
  } = useStore();

  let location = useLocation();
  let from = location.state?.from?.pathname || "/questions";

  function login(e) {
    e.preventDefault();
    loginThunk(form.username, form.password);
  }

  function onchange(e) {
    setForm({ ...form, [e.target.name]: e.target.value });
  }

  if (isAuth) return <Navigate to={from} replace />;
  return (
    <div style={{ width: 500, padding: 20 }}>
      <h1>Login</h1>
      <Form onSubmit={login}>
        <label htmlFor="email">Email</label>
        <Input
          onChange={onchange}
          name="username"
          value={form.username}
          type="text"
          id="email"
          required
        />
        <label htmlFor="password">Password</label>
        <Input
          type="password"
          name="password"
          onChange={onchange}
          id="password"
          value={form.password}
          required
        />
        <Button disabled={loading}>{loading ? "Processing" : "Login"}</Button>
      </Form>
    </div>
  );
}
