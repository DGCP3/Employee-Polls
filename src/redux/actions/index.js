export * as filterActions from "./filterActions";
export * as authActions from "./authActions";
export * as qaActions from "./Q&A_Actions";
