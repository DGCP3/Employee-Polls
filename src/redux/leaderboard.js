import { _getLeaderboard } from "../mock-api";

export default function leaderboardReducer(state = [], { type, payload }) {
  switch (type) {
    case "load_leaderboard":
      return [...payload];
    default:
      return state;
  }
}
const load = (payload) => ({ type: "load_leaderboard", payload });
export const loadLeaderboard = () => {
  return (dispatch) => {
    _getLeaderboard().then((res) => {
      const sorted = res.sort(
        (a, b) => b.answers + b.questions - (a.answers + a.questions)
      );
      dispatch(load(sorted));
    });
  };
};
