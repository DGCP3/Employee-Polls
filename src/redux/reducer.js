import { _createQuestion, _answerQuestion, _loginUser } from "../mock-api";
const SET_LOADING = "SET_LOADING";
const SET_USER = "SET_USER";
const SET_ERROR = "SET_ERROR";
const LOG_OUT = "LOG_OUT";
const SET_QUESTIONS = "SET_QUESTIONS";
const SET_ANSWERS = "SET_ANSWERS";

const initialState = {
  isAuth: false,
  loading: false,
  error: null,
  user: null,
  questions: [],
  answers: {},
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_LOADING:
      return {
        ...state,
        loading: true,
        error: null,
      };
    case SET_USER:
      return {
        ...state,
        loading: false,
        isAuth: true,
        user: action.payload,
      };
    case SET_ERROR:
      return {
        ...state,
        loading: false,
        error: action.payload,
      };
    case LOG_OUT:
      return {
        ...state,
        isAuth: false,
        user: null,
        questions: [],
        answers: {},
      };
    case SET_QUESTIONS:
      return {
        ...state,
        loading: false,
        error: null,
        questions: [...state.questions, ...action.payload],
      };
    case SET_ANSWERS:
      return {
        ...state,
        answers: {
          ...state.answers,
          ...action.payload,
        },
      };
    default:
      return state;
  }
};
// action creator functions
const setLoading = () => ({ type: SET_LOADING });
const loginSuccess = (payload) => ({ type: SET_USER, payload });
const loginFailure = (payload) => ({
  type: SET_ERROR,
  payload,
});
export const logout = () => ({
  type: LOG_OUT,
});

const setQuestions = (payload) => ({
  type: SET_QUESTIONS,
  payload,
});

const setAnswers = (payload) => ({
  type: SET_ANSWERS,
  payload,
});

// thunks
export const loginThunk = (username, password) => {
  return async (dispatch) => {
    dispatch(setLoading());
    try {
      const response = await _loginUser(username, password);
      if ("error" in response) throw new Error(response.error);
      const { answers, questions, ...rest } = response;
      dispatch(loginSuccess(rest));
      dispatch(setAnswers(answers));
      dispatch(setQuestions(questions));
    } catch (error) {
      dispatch(loginFailure(error.message));
    }
  };
};

export const createQuestionThunk = (options) => {
  return async (dispatch, getState) => {
    try {
      const {
        user: { id },
      } = getState();
      const response = await _createQuestion(options, id);
      dispatch(setQuestions([response.id]));
    } catch (error) {
      console.error(error);
    }
  };
};

export const setAnswersThunk = (_answer) => {
  return async (dispatch, getState) => {
    try {
      const {
        answers,
        user: { id },
      } = getState();
      const response = await _answerQuestion(
        id,
        {
          answers: { ...answers, ..._answer },
        },
        Object.keys(_answer)[0]
      );
      dispatch(setAnswers(response?.answers));
    } catch (error) {
      console.log(error);
    }
  };
};
