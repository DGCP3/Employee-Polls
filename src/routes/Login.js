export default function Login() {
  return (
    <div data-testId="loginForm">
      <div style={{ width: 500, padding: 20 }}>
        <h1>Login</h1>
        <form>
          <label htmlFor="email">Email</label>
          <input type="text" name="email" id="email" />
          <label htmlFor="password">Password</label>
          <input type="password" id="password" />
          <button>Login</button>
        </form>
      </div>
    </div>
  );
}
