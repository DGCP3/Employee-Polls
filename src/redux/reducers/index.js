import { combineReducers } from "redux";
import leaderboardReducer from "../leaderboard";
import authReducer from "./authReducer";
import filterReducer from "./filterReducer";
import QAreducer from "./Q&AReducer";

export default combineReducers({
  auth: authReducer,
  QA: QAreducer,
  filter: filterReducer,
  leaderboard: leaderboardReducer,
});
