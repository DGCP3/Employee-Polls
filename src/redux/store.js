import { applyMiddleware, compose, createStore } from "redux";
import thunk from "redux-thunk";
import { reducer } from "./reducer";

export const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  )
);
