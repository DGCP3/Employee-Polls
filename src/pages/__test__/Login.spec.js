import { render, cleanup, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { store } from "../../redux/store";
import Login from "../Login";

afterEach(cleanup);
describe("Login component", () => {
  it("Login", () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<Login />} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
    // snapshot
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("loginForm")).toBeInTheDocument();
    expect(screen.getByTestId("username")).toBeInTheDocument();
    expect(screen.getByTestId("password")).toBeInTheDocument();
    expect(screen.getByTestId("loginForm")).toHaveFormValues({
      username: "zoshikanlu",
      password: "pass246",
    });
  });
});
