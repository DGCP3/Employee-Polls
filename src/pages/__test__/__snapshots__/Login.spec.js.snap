// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login 1`] = `
<div>
  <div
    class="sc-kDDrLX flYHzi"
  >
    <form
      class="sc-iqcoie bDXuGN"
      data-testid="loginForm"
    >
      <label
        for="username"
      >
        Username:
      </label>
      <input
        class="sc-papXJ fixUxh"
        data-testid="username"
        id="email"
        name="username"
        required=""
        type="text"
        value="zoshikanlu"
      />
      <label
        for="password"
      >
        Password:
      </label>
      <input
        class="sc-papXJ fixUxh"
        data-testid="password"
        id="password"
        name="password"
        required=""
        type="password"
        value="pass246"
      />
      <button
        class="sc-gsnTZi jtOEId"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;

exports[`Login component Login 1`] = `
<div>
  <div
    class="sc-kDDrLX flYHzi"
  >
    <form
      class="sc-iqcoie bDXuGN"
      data-testid="loginForm"
    >
      <label
        for="username"
      >
        Username:
      </label>
      <input
        class="sc-papXJ fixUxh"
        data-testid="username"
        id="email"
        name="username"
        required=""
        type="text"
        value="zoshikanlu"
      />
      <label
        for="password"
      >
        Password:
      </label>
      <input
        class="sc-papXJ fixUxh"
        data-testid="password"
        id="password"
        name="password"
        required=""
        type="password"
        value="pass246"
      />
      <button
        class="sc-gsnTZi jtOEId"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
